test/cpp/util/spiffe_test_credentials.cc

```diff
@@ -0,0 +1,138 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include "test/cpp/util/spiffe_test_credentials.h"
+#include "test/core/end2end/data/ssl_test_data.h"
+
+namespace grpc {
+namespace testing {
+
+class TestSyncTlsCredentialReload
+    : public ::grpc_impl::experimental::TlsCredentialReloadInterface {
+  // Sync implementation.
+  int Schedule(
+      ::grpc_impl::experimental::TlsCredentialReloadArg* arg) override {
+    struct ::grpc_impl::experimental::TlsKeyMaterialsConfig::PemKeyCertPair
+        pem_key_cert_pair = {test_server1_key, test_server1_cert};
+    std::vector<
+        struct ::grpc_impl::experimental::TlsKeyMaterialsConfig::PemKeyCertPair>
+        pem_key_cert_pair_list = {pem_key_cert_pair};
+    arg->set_key_materials(test_root_cert, pem_key_cert_pair_list);
+    arg->set_status(GRPC_SSL_CERTIFICATE_CONFIG_RELOAD_NEW);
+    return 0;
+  }
+};
+
+class TestSyncTlsServerAuthorizationCheck
+    : public ::grpc_impl::experimental::TlsServerAuthorizationCheckInterface {
+  // Sync implementation.
+  int Schedule(
+      ::grpc_impl::experimental::TlsServerAuthorizationCheckArg* arg) override {
+    GPR_ASSERT(arg != nullptr);
+    arg->set_success(1);
+    arg->set_status(GRPC_STATUS_OK);
+    return 0;
+  }
+};
+
+static void TestAsyncTlsServerAuthorizationCheckCallback(
+    ::grpc_impl::experimental::TlsServerAuthorizationCheckArg* arg) {
+  GPR_ASSERT(arg != nullptr);
+  arg->set_success(1);
+  arg->set_status(GRPC_STATUS_OK);
+  arg->OnServerAuthorizationCheckDoneCallback();
+}
+
+class TestAsyncTlsServerAuthorizationCheck
+    : public ::grpc_impl::experimental::TlsServerAuthorizationCheckInterface {
+ public:
+  TestAsyncTlsServerAuthorizationCheck(SpiffeThreadList* thread_list)
+      : thread_list_(thread_list) {}
+
+  // Async implementation.
+  int Schedule(
+      ::grpc_impl::experimental::TlsServerAuthorizationCheckArg* arg) override {
+    GPR_ASSERT(arg != nullptr);
+    server_authz_check_thread_ =
+        std::thread(TestAsyncTlsServerAuthorizationCheckCallback, arg);
+    thread_list_->add_thread(std::move(server_authz_check_thread_));
+    return 1;
+  }
+
+ private:
+  std::thread server_authz_check_thread_;
+  SpiffeThreadList* thread_list_;
+};
+
+/** This method creates a TlsCredentialsOptions instance with no key materials,
+ *  whose credential reload config is configured using the
+ *  TestSyncTlsCredentialReload class, and whose server authorization check
+ *  config is determined as follows:
+ *  - if |is_client| is true,
+ *      -if |is_async|, configured by TestAsyncTlsServerAuthorizationCheck,
+ *      -otherwise, configured by TestSyncTlsServerAuthorizationCheck.
+ *  - otherwise, the server authorization check config is not populated.
+ *
+ *  Further, the cert request type of the options instance is always set to
```

Comment:
I think GRPC_SSL_REQUEST_AND_REQUIRE_CLIENT_CERTIFICATE_AND_VERIFY will only take effect at server side, not client side. Since client will always validate server's certificate.

Reply:
But I have to put some value for the client's cert request type, even if it doesn't do anything, right? For that reason, I just pass in `GRPC_SSL_REQUEST_AND_REQUIRE_CLIENT_CERTIFICATE_AND_VERIFY` for both. If you think there's a better "default" option though, let me know! :)

Reply:
For now you do not have to. But it may change in future (if we decide to support the feature that client also does not validate server's certificate).

Reply:
Sounds good. Thanks!